Sticky count redesign

braem opened the pull request:
Closes #917 
Closes #934 

Implemented a new design, where stickybombs are represented as boxes on the hud. Shooting triggers a simple color animation that represents arm time. Also includes a "pre-arm" color, which it changes to right when shot, then fades into the arm color.

Networked a bool that represents whether a sticky can explode (ie. it's not in a `func_nogrenades`). Used this to display the boxes as red when their particular stickies cannot explode.

#934 was caused by `AddStickybomb` not setting the networked count in the sticky launcher code. It is called from `OnDataChanged` in the stickybomb itself.

I found that `GetStickyByCount` in the sticky launcher crashed the game. I found this to only occur right when a sticky is fired, and when `m_iStickybombCount != m_Stickybombs.Size()`. So I added a check there that returns `nullptr` instead of hitting an access oob error. Not sure why this occurs nor how to fix that properly.

https://cdn.discordapp.com/attachments/634116622159708193/753501823469617213/2020-09-09_23-26-10.mp4
https://cdn.discordapp.com/attachments/634116622159708193/753501883947155517/2020-09-09_23-26-36.mp4

In these videos it is off-center, before I added the autolayout that helps with different resolutions:
![image](https://user-images.githubusercontent.com/9014762/92692369-e8d05500-f2f8-11ea-84a8-9a0f2c2ae226.png)

### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_stickybombs.cpp:
```diff
@@ -1,20 +1,29 @@
+    Color *color;
+    const char *animationSeq;
+    bool flag;
+};
```
This is a step in the right direction but this calls for a `StickyBox` vgui control. It doesn't seem to need to be a button, either. Buttons are inherently clickable, this is just a box that has its bg and fg colors animated.

The animation sequence would be the same, you would just be running it explicitly on the StickyBox panel itself.

braem replied:
Yeah I was using button as it deals with the border bg coloring but doesn't make much sense here. 
I understand a control for the animation and color sake, but not sure where to keep the flag I use. I'll stuff it in the control for now.

Gocnak replied:
Yes the goal is to convert `StickyBox` to be an actual control that has its own colors. Can "flag" be renamed? What does it even mean?

braem replied:
For the flag:
- `true` signifies that the sticky is new and can prearm (set prearm color and start the fade-in animation).
- `false` signifies that the stickybox has prearmed (started the animation) and can be reset to `NoStickyColor` when it is det.

So maybe a better name is `seen` or `processed`. 

It is tied to the control yes but it still needs to be accessed by the parent `hud_stickybombs` as the animations are swapped when a sticky is det. This is to keep their current animation when they are moved leftward.
This animation swapping i think is the only thing that really needs to be in the parent, as far as `OnThink` is concerned.

Gocnak commented on mp/src/game/shared/momentum/mom_stickybomb.cpp:
```diff
@@ -325,4 +328,11 @@ void CMomStickybomb::VPhysicsCollision(int index, gamevcollisionevent_t *pEvent)
+{
+    BaseClass::VPhysicsUpdate(pPhysics);
+    m_bCanExplode.Set(!CNoGrenadesZone::IsInsideNoGrenadesZone(this) && !m_bFizzle);
+}
```
Can this instead be done inside the nogrenades trigger `OnStartTouch`/`OnEndTouch`? You may have to play around with the entity collision flags to get them to trigger things but I think this may be a better way than having every sticky constantly for-loop the triggers.

braem replied:
Sure can :)

braem commented:
Moved functionality to a `Stickybox` control. This has a `Process` function that takes in a process type. Changed the `flag` to a member variable `m_bProcessed`. Moved the different colors into the stickyboxes themselves. Also unpinned from the charge meter as it seems appropriate; other larger hud elements don't seem to be pinned.

Gocnak commented on mp/game/momentum/resource/ui/HudStickybombs.res:
```diff
@@ -2,37 +2,51 @@
+        "tall" "0" // see "BoxDimension" above
+        "NoStickyColor" "BlackHO"
+        "PreArmColor"   "BlackHO"
+        "DisabledColor" "MomentumRed"
```
One of the things I'm trying to push towards is not having duplicated logic or resources, think of the components inside our frontend website code.

We have to think about what exactly the role of StickyBox is. A StickyBox is a panel that is rendered as a square that gets its background color animated to match the state of the associated stickybomb from the launcher. This means that its state is going to be managed entirely by the hud_stickybombs class. Think of the speedometer and its speedometer labels. The HUD element manages all the labels, but the labels given the state handle all the rendering and animation. That's what we should be trying to go for here.

With this in mind, we can start to think about how best to define StickyBox. The HUD element is going to be constantly checking the status of the stickybombs and setting internal state of each of these StickyBoxes. But the catch is we want every single StickyBox to behave the exact same given the same internal state. I know there's swapping, but why not just swap the location of the panel instead of internal states?

The goal is to get rid of the idea of three distinct controls being defined and animated etc. They should all have the same panel name, have their settings handed to them by the hud panel, and all the res files shouldn't contain duplicates. Keep in mind you can always define colors in the scheme for use in specific panels, which are then used in animations as well, if need be. Another thing to figure out is running an animation on an individual control, as we'd prefer that to happen compared to three animations animating three different state variables in the HUD class. Think of the StickyBox as tied to a particular stickybomb.

Gocnak commented on mp/game/momentum/scripts/HudAnimations.txt:
```diff
@@ -113,6 +113,19 @@ event FadeOutStageVel
 
+event FirstStickyArm
+{
+    Animate CHudStickybombs FirstStickyColor "MomentumBlue" Accel 0.01 0.55
```
Another thing to consider is you should be able to reference CPanelAnimationVars here as the target color as well. So this could be "StickyArmColor" as an animation var reference and lets users definite a different arm color via the res file.

braem replied:
I was not able to reference it as the target. Looking at the other animations I think you can only use ones in `SourceScheme.res`, or a literal.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_stickybombs.cpp:
```diff
@@ -67,19 +76,54 @@ void CHudStickybombs::OnThink()
+        }
+        else
+        {
+            m_Stickyboxes[i]->Process(STICKYBOX_PROCTYPE_NOSTICKY);
```
I'm not sure I understand why process takes an enum. They're all pretty distinct calls with no overlapping, so why not just call the individual process functions here?

Gocnak commented on mp/src/game/client/momentum/ui/controls/Stickybox.h:
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <vgui_controls/EditablePanel.h>
```
It doesn't use any res file itself and can therefore just be a Panel

Gocnak commented on mp/src/game/server/momentum/mom_triggers.cpp:
```diff
@@ -1557,6 +1557,30 @@ void CNoGrenadesZone::Precache()
 
+void CNoGrenadesZone::OnStartTouch(CBaseEntity* pOther)
+{
+    if(!pOther)
```
Spacing

Gocnak commented on mp/src/game/client/momentum/ui/controls/Stickybox.cpp:
```diff
@@ -0,0 +1,90 @@
+    if (!m_pszAnimationName || !m_pAnimationColor)
+        return false;
+
+    return g_pClientMode->GetViewportAnimationController()->StopAnimationSequence(GetParent(), m_pszAnimationName);
```
Does just using `vgui::AnimationController()` here work?

braem replied:
From what I see the animation controller is created in `baseviewport` which we access via the clientmode global. Is there some other way to get that that I'm missing? I did this similar to how I did it for `SpeedometerLabel`.

braem commented:
Fixed most of the issues brought up and made the commits nice. I changed that "processed" flag to `m_bAnimationStarted` as it correlates with when the animation is start/stopped.

Added a border, same as the progress bar as the charge meter and this go together:
![image](https://user-images.githubusercontent.com/9014762/93568343-dda6a600-f944-11ea-9437-5e430ce68546.png)

The design now scales with the stickybomb count:
![image](https://user-images.githubusercontent.com/9014762/93568597-40983d00-f945-11ea-82f8-04fd325e3f82.png)

The biggest question mark for me is in the animation. It doesn't work right if I put it in `HudAnimations.txt`, as I think that is linked to the class statically? For now I am manually running an animation command, using the arm time for stickybombs (which I made client facing via static getter to do so):
`g_pClientMode->GetViewportAnimationController()->RunAnimationCommand(this, "StickyColor", target, 0.0f, C_MomStickybomb::GetArmTime(), AnimationController::INTERPOLATOR_ACCEL);`

Is there a better way of doing this?

Gocnak commented on mp/src/game/server/momentum/mom_triggers.cpp:
```diff
@@ -1557,6 +1557,30 @@ void CNoGrenadesZone::Precache()
     PrecacheModel(NOGRENADE_SPRITE);
 }
 
+void CNoGrenadesZone::OnStartTouch(CBaseEntity* pOther)
+{
+    if (!pOther)
+        return;
+
+    const auto pSticky = dynamic_cast<CMomStickybomb *>(pOther);
+    if (!pSticky)
+        return;
+
+    pSticky->SetCanExplode(false);
+}
+
+void CNoGrenadesZone::OnEndTouch(CBaseEntity* pOther)
+{
+    if (!pOther)
+        return;
+
+    const auto pSticky = dynamic_cast<CMomStickybomb *>(pOther);
+    if (!pSticky)
+        return;
+
+    pSticky->SetCanExplode(true);
+}
```
Always call your base class methods.

braem replied:
Ah I glanced at these and figured I didn't have to as it was prefixed "On", woops

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_stickybombs.cpp:
```diff
@@ -67,19 +120,66 @@ void CHudStickybombs::OnThink()
     {
-        m_pStickyLabel->SetVisible(true);
+        m_Stickyboxes[i]->SetBgColor(m_NoStickyColor);
+        m_Stickyboxes[i]->SetBorder(pScheme->GetBorder("ButtonDepressedBorder"));
```
You can alias a border to another one in the scheme. I would recommend doing something like:

```suggestion
        m_Stickyboxes[i]->SetBorder(pScheme->GetBorder("StickyboxBorder"));
```

And then in `ClientScheme.res`:

```
. . .

Borders
{
    . . .

    StickyboxBorder ButtonDepressedBorder

    . . .
}
```

braem replied:
I tried this but one thing I noticed is that `ClientScheme.res` doesn't seem to be able to reference values in `SourceScheme.res` like this. I have to place this in either `SourceScheme.res` or in the "BaseSettings" section of `ClientScheme.res`.

Also noticed this when I was trying to change the speedometer increase/decrease/normal colors to "MomentumBlue"/"MomentumRed"/etc.

Lastly, doing it this way makes the borders a different color for some reason:
![image](https://user-images.githubusercontent.com/9014762/93661570-9cc69400-fa0d-11ea-88c7-d31b8828ec44.png)
s o r s e

braem replied:
Nevermind I figured this out. The border shown above is the default one when `pScheme->GetBorder` doesn't find it.

I put it in the border section and had it alias `FrameBorder` instead of `ButtonDepressedBorder`. I'm guessing because `ButtonDepressedBorder` is an alias of `FrameBorder` it just doesn't work? Weird.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_stickybombs.cpp:
```diff
@@ -1,20 +1,62 @@
+void Stickybox::StartAnimation(Color start, Color target)
+{
+    m_StickyColor = start;
+    g_pClientMode->GetViewportAnimationController()->RunAnimationCommand(this, "StickyColor", target, 0.0f, C_MomStickybomb::GetArmTime(), AnimationController::INTERPOLATOR_ACCEL);
```
Regarding your [comment](https://github.com/momentum-mod/game/pull/935#issuecomment-694706920):
> Is there a better way of doing this?

Yes, but involves getting a little bit dirty.

AnimationController has only two ways to run animations, `StartAnimationSequence` and `RunAnimationCommand`. What we want is a hybrid between the two; running a specific sequence, but for a specific panel, while ignoring the actual panel name defined in the animation. I'm genuinely surprised this isn't possible, as `StartAnimationSequence` has an overload definition that takes a panel, but is supposed to be the *starting point for finding the named panel in the sequence*. Thanks Valve. 🙄 

So what I want you to do is make a `StartAnimationSequenceForPanel(Panel *pPanel, const char *pSequence)` method. This will effectively do the best of both worlds, so it'll look like:
```cpp
bool AnimationController::StartAnimationSequenceForPanel(Panel *pPanel, const char *sequenceName)
{
    Assert( pPanel);

	if (m_bAutoReloadScript)
	{
		// Reload the script files
		ReloadScriptFile();
	}

	// lookup the symbol for the name
	UtlSymId_t seqName = g_ScriptSymbols.Find(sequenceName);
	if (seqName == UTL_INVAL_SYMBOL)
		return false;

	// Msg("Starting animation sequence %s\n", sequenceName);

	// remove the existing command from the queue
	RemoveQueuedAnimationCommandsForPanel(seqName, pPanel);

	// look through for the sequence
	int i;
	for (i = 0; i < m_Sequences.Count(); i++)
	{
		if (m_Sequences[i].name == seqName)
			break;
	}
	if (i >= m_Sequences.Count())
		return false;

	// execute the sequence
	for (int cmdIndex = 0; cmdIndex < m_Sequences[i].cmdList.Count(); cmdIndex++)
	{
		ExecAnimationCommandForPanel(seqName, m_Sequences[i].cmdList[cmdIndex], pPanel);
	}

	return true;	
}

```

The goal here is to make it so that instead of *searching* for the target panel (which is typically defined in the HudAnimations.txt file), we're just *giving* it to the methods. So then you'll need to copy the implementation of `ExecAnimationCommand` and make it automatically use the `StartCmd_Animate` that can directly use the panel pointer of `pPanel`.

You may also need to define a `RemoveQueuedAnimationCommandsForPanel` that does the same as `RemoveQueuedAnimationCommands` but for a given panel (see how it searches for the target panel again?).

Gocnak replied:
Oh and try to use `vgui::GetAnimationController` instead of having to fetch it from the clientmode. If animations break, feel free to use the clientmode one, but for the most part, the vgui namespace one should be good to use (and less verbose!).

braem commented:
Implemented the requested changes, apart from two relating to colors/borders ( https://github.com/momentum-mod/game/pull/935#discussion_r491351840 and https://github.com/momentum-mod/game/pull/935#discussion_r491308863 )

braem commented:
Rebased & figured out the border stuffs. Unsure if there is a way to reference an animation var or color in `ClientScheme.res` in `HudAnimations.txt` as the target now.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_stickybombs.cpp:
```diff
@@ -1,20 +1,64 @@
+void Stickybox::StartAnimation(Color startColor)
+{
+    m_StickyColor = startColor;
+    vgui::GetAnimationController()->StartAnimationSequenceForPanel(this, STICKY_COUNT_ARM_ANIM_NAME);
```
You are using namespace vgui so you can remove `vgui::` from this.

braem replied:
i took that change request a little too literally 😑

Gocnak commented on mp/src/vgui2/vgui_controls/AnimationController.cpp:
```diff
@@ -1023,10 +1063,25 @@ bool AnimationController::StopAnimationSequence( Panel *pWithinParent, const cha
 // Purpose: Runs a custom command from code, not from a script file
 //-----------------------------------------------------------------------------
-void AnimationController::CancelAnimationsForPanel( Panel *pWithinParent )
+void AnimationController::CancelAnimations( Panel *pWithinParent )
```
You now need to verify things are calling the right function. The old cancel method was checking the parent and finding the child, but since you renamed this to something else, was there something else in the code calling the old function name? The new function you added took its old name and does slightly different behavior; this could be a regression if something was using the old method.

braem replied:
I went through the code first and ensured nothing was calling this function before making this change.
